pkg/lightning/backend/tidb/tidb.go

```diff
@@ -364,31 +378,99 @@ func (be *tidbBackend) ImportEngine(context.Context, uuid.UUID) error {
 }
 
 func (be *tidbBackend) WriteRows(ctx context.Context, _ uuid.UUID, tableName string, columnNames []string, rows kv.Rows) error {
-	var err error
-outside:
+rowLoop:
 	for _, r := range rows.SplitIntoChunks(be.MaxChunkSize()) {
+		batch := true
+		rowsToSkip := make(map[int]struct{})
+	retryLoop:
 		for i := 0; i < writeRowsMaxRetryTimes; i++ {
-			err = be.WriteRowsToDB(ctx, tableName, columnNames, r)
+			errRowIdx, err := be.WriteRowsToDB(ctx, tableName, columnNames, r, batch, rowsToSkip)
 			switch {
 			case err == nil:
-				continue outside
+				continue rowLoop
 			case common.IsRetryableError(err):
 				// retry next loop
 			default:
-				return err
+				// retry next loop when batch is true, which means the last WriteRowsToDB failed in the batch mode,
+				// we need to redo the writing row-by-row to find where the error located.
+				if batch {
+					i = 0
+					batch = false
+					continue retryLoop
+				}
+				// If batch is false, it means we find the row where the non-retryable error located,
+				// now we should record and skip it if necessary.
+				be.curErrorCount++
+				if be.curErrorCount >= be.maxErrorCount {
+					return errors.Annotatef(err, "[%s] write rows reach max error count %d", tableName, be.maxErrorCount)
+				}
+				// TODO: record the error.
+				// Reset the retry count before we skip the error and continue inserting the rest of rows.
+				i = 0
+				// Update the row we should skip.
+				rowsToSkip[errRowIdx] = struct{}{}
 			}
 		}
-		return errors.Annotatef(err, "[%s] write rows reach max retry %d and still failed", tableName, writeRowsMaxRetryTimes)
 	}
 	return nil
 }
 
-func (be *tidbBackend) WriteRowsToDB(ctx context.Context, tableName string, columnNames []string, r kv.Rows) error {
+type stmtTask struct {
+	rowIdx int
+	stmt   string
+}
+
+// If batch is true, we will insert multiple rows like this:
+//   start transaction;
+//   insert into t1 values (111), (222), (333), (444);
+//   commit;
+// If batch is false, we will insert multiple rows like this:
+//   start transaction;
+//   insert into t1 values (111);
+//   insert into t1 values (222);
+//   insert into t1 values (333);
+//   insert into t1 values (444);
+//   commit;
+// See more details in br#1366: https://github.com/pingcap/br/issues/1366
+// WriteRowsToDB will return the index of error row and its corresponding error if any error occurs.
+func (be *tidbBackend) WriteRowsToDB(ctx context.Context, tableName string, columnNames []string, r kv.Rows, batch bool, rowsToSkip map[int]struct{}) (int, error) {
 	rows := r.(tidbRows)
 	if len(rows) == 0 {
-		return nil
+		return 0, nil
 	}
 
+	insertStmt := be.prepareStmt(tableName, columnNames)
+	// Note: we are not going to do interpolation (prepared statements) to avoid
+	// complication arise from data length overflow of BIT and BINARY columns
+	var stmtTasks []stmtTask
+	if batch {
```

Comment:
I think the bool is still needed here because the rows we pass are the same and we can only tell whether it's a batch mode by this flag according to the current implementation.

Reply:
If we convert the rows to tidbRows in the caller, we can also take slice of the rows to avoid introduce startIndex and endIndex